File: src/ai/metrics/ModelEvaluator.java
package ai.metrics;

import java.util.*;


public class ModelEvaluator
{
  public static double getAccuracy(double[]yLabels, double[]predictedYLabels)
  {
    double good = 0;
    double bad = 0;
    double accuracy = 0;

    for (int i = 0; i < predictedYLabels.length; i++)
    {
      if (yLabels[i] == predictedYLabels[i])
      {
        good++;
        //System.out.println(yLabels[i] + "  good   "  + predictedYLabels[i]);
      }
      else
      {
        bad++;
        //System.out.println(yLabels[i] +"  bad   "  + predictedYLabels[i]);
      }
    }




    accuracy = (good/(good+bad))*100;
    return accuracy;
  }

  public static double getBaselineAcc(double[] yLabels)
  {
    int one = 0;
    int zero = 0;
    double mostfreq = 0;

    for (int i = 0; i < yLabels.length; i++)
    {
      if(yLabels[i] == 1.0)
      {
        one++;
      }
      else
      {
        zero++;
      }
    }

    if (one > zero)
    {
      mostfreq = one;
    }
    else
    {
      mostfreq = zero;
    }

    return mostfreq/(one+zero);

  }

  //c.m. for a binary situation. everything should be 0 or 1.
  public static void confusionMatrix(double[] yLabels, double[] predictedYLabels)
  {
    int tp = 0; //true positives
    int tn = 0; //true negatives
    int fp = 0; //false positives
    int fn = 0; //false negatives
    int actualZero = 0;
    int actualOne = 0;
    int predictedOne = 0;

    for(int i = 0; i < predictedYLabels.length; i++)
    {
      //true positive
      if ((yLabels[i] == predictedYLabels[i]) && (yLabels[i] == 1))
      {
        tp++;
        actualOne++;
        predictedOne++;
      }
      //true negatives
      else if ((yLabels[i] == predictedYLabels[i]) && (yLabels[i] == 0))
      {
        tn++;
        actualZero++;
      }
      //false positives
      else if ((yLabels[i] == 0) && (predictedYLabels[i] == 1))
      {
        fp++;
        actualZero++;
        predictedOne++;
      }
      //false negatives
      else if ((yLabels[i] == 1) && (predictedYLabels[i] == 0))
      {
        fn++;
        actualOne++;
      }
    }

    System.out.print("\n*****Confusion Matrix*****\n");
    System.out.print("\tPredicted:0\tPredicted:1\n");
    System.out.print("Actual:0\t" + tn +"\t" + fp + "\n");
    System.out.print("Actual:1\t" + fn +"\t" + tp + "\n");
    System.out.println("\nAccuracy =\t\t" + ((double)(tp+tn)/(tp+tn+fp+fn)));
    System.out.println("Error Rate =\t\t" + ((double)(fp+fn)/(tp+tn+fp+fn)));
    System.out.println("Sensitivity/Recall =\t" + ((double)tp/actualOne));
    System.out.println("False Positive Rate =\t" + ((double)fp/actualZero));
    System.out.println("Specificity =\t\t" + ((double)tn/actualZero));
    System.out.println("Precision =\t\t" + ((double)tp/predictedOne));
    System.out.println("Prevalence =\t\t" + ((double)actualOne/(tp+tn+fp+fn)));

  }

  public static double linearSum(double[][] x, double[] beta, int row)
  {
      double sum = 0.;

      sum = sum + beta[0];

      for (int i = 1; i < beta.length; i++)
      {
          sum = sum + (beta[i] * x[row][i-1]);
      }

      return sum;
  }

  public static double hypothesis(double[][] x, double[] beta, int row)
  {
      double sigmoid = 0.;

      sigmoid = 1/(1+Math.pow(Math.E,-linearSum(x, beta, row)));

      return sigmoid;
  }

  public static double logLikelihood(double[][] x, double[] y, double[] beta)
  {
    int m = x.length;
    double cost = 0.;
    double sum = 0.;
    int n = beta.length;

    for (int i = 0; i < m; i++)
    {
        if (y[i] == 1.0)
        {
          sum = sum + (Math.log(hypothesis(x, beta, i)));
        }
        else
        {
          sum = sum + Math.log(1 - hypothesis(x, beta, i));
        }
    }

    cost= (sum/m);

    return cost;
  }

  public static double mcfaddenRSquared(double[][] x, double[] yLabels, double[] beta)
  {
    double rSquare = 0;
    double ones = 0;
    double zeroes = 0;
    double overallProb = 0;
    double[] overallBeta = new double[1];
    double logLikeliOverall = 0;
    double logLikeliModel = 0;

    for (int i = 0; i < yLabels.length; i++)
    {
      if (yLabels[i] == 1)
      {
        ones++;
      }
      else
      {
        zeroes++;
      }
    }

    overallProb = ones/(zeroes+ones);
    overallBeta[0] = overallProb;

    logLikeliOverall = logLikelihood(x, yLabels, overallBeta);
    logLikeliModel = logLikelihood(x, yLabels, beta);

    rSquare = (logLikeliOverall - logLikeliModel) / logLikeliOverall;



    return rSquare;
  }

  public static void rankWeights(double[] array)
  {
    double temp = 0;
    double max = 0.;
    int maxIndex = 0;
    int count = 0;

    double[] original = new double[array.length];

    for (int i = 0; i < original.length; i++)
    {
      original[i] = array[i];
    }

    System.out.println("\nWeight Ranking\nCoefficient\t\tFeatureNumber");


		for (int j = 0; j < array.length; j++)
		{
			max = array[count];

			for (int i = count; i < array.length; i++)
			{
				if (Math.abs(array[i]) >= Math.abs(max))
				{
					max = array[i];
					maxIndex = i;
				}
			}
			//System.out.print(max + "\t" + maxIndex + "\n");
			temp = array[count];
			array[count] = array[maxIndex];
			array[maxIndex] = temp;

			count++;
    }

    for(int i = 0; i < array.length; i++)
    {
      System.out.print(array[i] + "\t" + findOccurence(original, array[i]) + "\n");
    }


  }

  public static int findOccurence(double[] array, double search)
  {
    for (int i = 0; i < array.length; i++)
    {
      if(array[i] == search)
      {
        return i;
      }
    }
    return -1;
  }
}

File: CE/Jason/ModelEvaluator.js
//class ModelEvaluator{
    function getAccuracy(labels, predictedLabels){
        var good = 0;
        var total = 0;
        var accuracy = 0;
    
        for(var i = 0; i < predictedLabels.length; i++){
            if (labels[i] === predictedLabels[i]){
                good++;
                total++;
            } else {
                total++;
            }
        }
        accuracy = (good / total) * 100;
        return accuracy;
    }

    function confusionMatrix(labels, predictedLabels){
        var truePos = 0;    //true positives
        var trueNeg = 0;    //true negatives
        var falPos = 0;     //false positives
        var falNeg = 0;     //false negatives
        var actual1 = 0;
        var actual0 = 0;
        var predicted1= 0;
    
        for(var i = 0; i < predictedLabels.length; i++){
            //true positive
            if(labels[i] === predictedLabels[i] && labels[i] === 1){
                truePos++;
                actual1++;
                predicted1++;
            //true negative
            } else if (labels[i] === predictedLabels[i] && labels === 0){
                trueNeg++;
                actual0++;
            //false positive
            } else if (labels[i] === 0 && predictedLabels[i] === 1){
                falPos++;
                actual0++;
                predicted1++;
            //false negative
            } else if (labels[i] === 1 && predictedLabels[i] === 0){
                falNeg++;
                actual1++;
            }
            console.log("*****Confusion Matrix*****");
            console.log("\tPredicted:0\tPredicted:1");
            console.log("Actual:0\t" + trueNeg +"\t" + falPos);
            console.log("Actual:1\t" + falNeg +"\t" + truePos);
            console.log("Accuracy =\t\t" + ((truePos + trueNeg) / (truePos + trueNeg + falPos + falNeg)));
            console.log("Error Rate =\t\t" + ((falPos + falNeg) / (truePos + trueNeg + falPos + falNeg)));
            console.log("Sensitivity/Recall =\t" + (truePos / actual1));
            console.log("False Positive Rate =\t" + (falPos / actual0));
            console.log("Specificity =\t\t" + (trueNeg / actual0));
            console.log("Precision =\t\t" + (truePos / predicted1));
            console.log("Prevalence =\t\t" + (actual1 / (truePos + trueNeg + falPos + falNeg)));
        }
    }

    function getBaselineAcc(labels){
        var one = 0;
        var zero = 0;
        var mostFreq = 0;
    
        for(var i = 0; i < labels.length; i++){
            if(labels[i] === 1){
                one++;
            } else {
                zero++;
            }
        }
    
        if (one > zero) {
          mostFreq = one;
        } else {
          mostFreq = zero;
        }
    
        return mostFreq/(one+zero);
    }

    function linearSum(x, beta, row) {
        var sum = 0.;
    
        sum = sum + beta[0];
    
        for (var i = 1; i < beta.length; i++) {
            sum = sum + (beta[i] * x[row][i-1]);
        }
    
        return sum;
    }

    function hypothesis(x, beta, row) {
        var sigmoid = 0.;
        sigmoid = 1/1+(Math.pow(Math.E,-linearSum(x,beta,row)));
        return sigmoid;
    }

    function logLikelihood(x, y, beta){
        var m = x.length;
        var cost = 0;
        var sum = 0;
        var n = beta.length;
    
        for(var i = 0; i < m; i++) {
            if(y[i] === 1.0){
                sum = sum + (Math.log(hypothesis(x, beta, i)));
            } else {
                sum = sum + Math.log(1 - hypothesis(x, beta, i));
            }
        }
    
        cost = sum/m;
        return cost;
    }

    function mcFaddenRSquared(x, labels, beta){
        var rSquare = 0;
        var ones = 0;
        var zeroes = 0;
        var overallProb = 0;
        var overallBeta = new Array(1);
        var logLikeliOverall = 0;
        var logLikeliModel = 0;
    
        for (var i = 0; i < yLabels.length; i++) {
            if (labels[i] == 1) {
                ones++;
            } else {
                zeroes++;
            }
        }
    
        overallProb = ones/(zeroes+ones);
        overallBeta[0] = overallProb;
    
        logLikeliOverall = logLikelihood(x, yLabels, overallBeta);
        logLikeliModel = logLikelihood(x, yLabels, beta);
    
        rSquare = (logLikeliOverall - logLikeliModel) / logLikeliOverall;
    
        return rSquare;
    }
    
    function rankWeights(array){
        var temp = 0;
        var max = 0.;
        var maxIndex = 0;
        var count = 0;
        var original = new Array(array.length);
    
        for (var i = 0; i < original.length; i++)
        {
          original[i] = array[i];
        }
    
        console.log("Weight Ranking\nCoefficient\t\tFeatureNumber");
    
    
            for (var i = 0; i < array.length; i++) {
                max = array[count];
    
                for (var j = count; j < array.length; j++) {
                    if (Math.abs(array[j]) >= Math.abs(max)) {
                        max = array[j];
                        maxIndex = j;
                    }
                }
                //console.log(max + "\t" + maxIndex);
                temp = array[count];
                array[count] = array[maxIndex];
                array[maxIndex] = temp;
    
                count++;
        }
    
        for(var i = 0; i < array.length; i++) {
          console.log(array[i] + "\t" + findOccurence(original, array[i]));
        }
    }
    
    function findOccurence(array, num){
        for(var i = 0; i < array.length; i++){
            if (array[i] === num){
                return i;
            }
        }
        return -1;
    }
//}